### ISSUE TITLE
FPN-18 | Phase 4 | Optimization Pass (Model, Pipeline, and API)

### ISSUE THREAD

**Issue by LavanyaSrinivas06** [LavanyaSrinivas06/AI-First-Preauth-Fraud-Detection#67]
### Ticket ID: FPN-18
Phase: Phase 4 — Testing & Optimization
Branch Name: feature/FPN-18-optimizations

---

### Context
Based on FPN-17 results, reduce end-to-end latency and improve throughput.  
Typical hotspots: preprocessing overhead, model inference speed, JSON (de)serialization, and Python GIL contention.

---

### Scope
- Optimize preprocessing and model inference path.
- Reduce serialization overhead and response size.
- Enable efficient concurrency at the server level.
- Re-measure improvements and update performance report.

---

### Steps

1. **Preprocessing optimizations**
   - Cache feature order and encoder transforms; avoid repeated allocations.
   - Convert dataframe → numpy early; minimize conversions.
   - If feasible, precompute constant features and prune unused columns.

2. **Model-level optimizations**
   - XGBoost:
     - Save and load with `save_model`/`load_model` if faster.
     - Consider `nthread` tuning to match CPU cores.
   - Autoencoder:
     - Enable TensorFlow inference optimizations (`tf.config.optimizer.set_jit(True)` where applicable).
     - Batch single-record requests into micro-batches (e.g., 8–16) if SLA allows.

3. **Numerical libraries and BLAS**
   - Ensure optimized BLAS (OpenBLAS/MKL) is used by numpy/xgboost.
   - Pin versions known to be fast in `requirements.txt`.

4. **Serialization and payload**
   - Use orjson/uvicorn httptools for faster JSON:
     ```bash
     uvicorn api.main:app --workers 4 --loop uvloop --http httptools
     ```
   - Switch FastAPI JSON response to orjson:
     ```python
     app = FastAPI(default_response_class=ORJSONResponse)
     ```
   - Trim response fields to essentials; avoid huge arrays.

5. **Server concurrency**
   - Select optimal `--workers` based on CPU cores and model thread use.
   - Investigate async request handling if IO-bound (e.g., remote feature stores later).

6. **Caching and warmup**
   - Warm up models on startup with a dummy request.
   - Cache SHAP reason codes for repeated requests (optional if needed later).

7. **Re-test and quantify gains**
   - Re-run FPN-17 scenarios with the same parameters.
   - Update `docs/performance_report.md` with before/after tables and graphs.
   - Note any trade-offs (e.g., micro-batching adds small queueing latency).

8. **Guardrails**
   - Add simple circuit-breaker or overload protection to return 503 when queue depth exceeds threshold (optional).

---

### Acceptance Criteria
- [ ] Measurable latency reduction vs. FPN-17 baseline (report p50/p90/p99 deltas).
- [ ] Throughput increases or remains stable with lower latency.
- [ ] No correctness regressions in unit/functional tests.
- [ ] Performance report updated with optimization details and final numbers.

---

### Artifacts
- Updated `api/main.py`, `api/service.py` (orjson, warmup)
- Updated `requirements.txt` (orjson, uvloop/httptools if used)
- Updated `docs/performance_report.md` with before/after results

---

### Helpful Links
- orjson FastAPI integration: https://fastapi.tiangolo.com/advanced/custom-response/#orjsonresponse
- TensorFlow XLA/JIT: https://www.tensorflow.org/xla
- XGBoost performance tips: https://xgboost.readthedocs.io/en/stable/tutorials/param_tuning.html

---

### Effort Estimate
M (1–2 days)
